File: interface/src/com/christopher/robot/message/Response.java
package com.christopher.robot.message;

import java.nio.ByteBuffer;

public abstract class Response
{
    protected MessageHeader header;
    protected MessageTrailer trailer;

    public abstract void pack( ByteBuffer buffer ); 
    //public abstract Response unpack( ByteBuffer buffer );

    protected Response()
    {
    }

    protected Response( int dataLength, int type, int id, int sequenceId,
                        int version )
    {
        header = new MessageHeader( MessageHeader.getSize() + dataLength + MessageTrailer.getSize(),
                                    type, id, sequenceId, version );
        trailer = new MessageTrailer();
    }

    protected Response( int dataLength, int type, int id, int version )
    {
        header = new MessageHeader( MessageHeader.getSize() + dataLength + MessageTrailer.getSize(),
                                    type, id, version );
        trailer = new MessageTrailer();
    }

    public MessageHeader getHeader()
    {
        return header;
    }

    public MessageTrailer trailer()
    {
        return trailer;
    }
}

File: simulator/src/message/Response.cpp

#include "Response.h"

Response::Response()
{
    header  = nullptr;
    trailer = nullptr;
}

Response::Response( int dataLength, int type, int id, int sequenceId, int version )
{
    header = new MessageHeader( MessageHeader::getHeaderSize() + dataLength
                              + MessageTrailer::getTrailerSize(), type, id,
                                sequenceId, version );
    trailer = new MessageTrailer();
}

Response::Response( int dataLength, int type, int id, int version )
{
    header = new MessageHeader( MessageHeader::getHeaderSize() + dataLength
                              + MessageTrailer::getTrailerSize(), type, id,
                                version );
    trailer = new MessageTrailer();
}

Response::~Response()
{
    if( header != nullptr )
    {
        delete header;
    }

    if( trailer != nullptr )
    {
        delete trailer;
    }
}

MessageHeader *
Response::getHeader()
{
    return header;
}

MessageTrailer *
Response::getTrailer()
{
    return trailer;
}
